### PR TITLE
Expand EXT_feature_metadata support

### PR DESCRIPTION
**Pull request by nithinp7**
Resolves: #513 
Addresses: #389 
Partially implements #771 

Expands EXT_feature_metadata support:
- Adds Blueprint-based "picking" to query metadata:
  - [x] Support querying Feature Id Textures.
  - [x] Support querying Feature Textures.
  - ***Note that vertex attributes picking has already previously been implemented.***

- Adds material graph based GPU "styling" where the metadata values can be used to drive material/shader effects:
  - [x] Support Feature Id Textures styling.
  - [x] Support Feature Textures styling.
  - [x] Support vertex attributes styling.

- Adds the `UCesiumEncodedMetadataComponent`, which can be added to tilesets to describe exactly what metadata to encode for availability for material styling. Specific properties out of specific feature tables can be picked out for gpu encoding, while others that aren't needed can be left out.

- Adds button on the `CesiumEncodedMetadataComponent` to auto-generate an incomplete material layer based on the currently selected metadata to encode. This generated material will take care of all the boiler plate code responsible for parsing feature id textures / attributes and performing the per-pixel / per-vertex feature id lookup. The requested property values will be available as the output pins of a custom shader node as plain floats or vectors.

Some Remaining TODOs:
- [x] Fix example materials, make them compatible for UE 4.26.
- [x] Fix picking blueprint in sample level + implement backward compatibility for blueprint api changes.
- [x] Easier view of available metadata on the `UCesiumEncodedMetadataComponent`, currently it is painful to manually open the tree.
- [x] Add an "Auto Fill" button to the encoded metadata component, that can automatically punch in all unique feature tables and properties (from the currently visible tiles). Users can do this to quickly get the superset of encodable properties and delete the ones they don't need. This might be particularly useful when users don't have convenient access to the glTF jsons or may have forgotten the exact names of the feature tables and properties in the metadata. 
- [ ] Check that picking and styling of metadata works on UE5 preview 1. May need to write Chaos-specific code to enable UV-based picking.
- [x] FIXED NOW (I think). There seems to be a strange issue with deletion of auto-generated materials causing crashes in some cases.
- [x] Add auto-generated materials for feature textures. 
- [ ] Occasional issues regarding using vertex interpolators in materials with multiple layers underneath. E.g., vertex interpolator in the metadata layer causes the material to not compile, if there is a base glTF layer. 

**Feature ID Textures queried in Blueprint using a line-trace.**
feature-id-textures3

**Feature ID Textures used to color the model based on age of components.**

**Per-Vertex Feature ID Attributes, in this case building height information, used to color the buildings of NYC**

**Blueprint-based metadata picking, used to spawn billboards indicating terrain-type.**

Example of how to use this:

##### After using "Auto Fill" the `CesiumEncodedMetadataComponent` should look something like this, listing the existing metadata in the tileset.
autoFill

##### After picking the required metadata properties and pressing "Generate Material", a material layer accessing the required metadata is created.
autoFill1

##### The output pins of the feature table lookup node can be used to access the metadata property values. They can be used to drive any shading effect. Here we highlight buildings built in 1931. Note that for vertex attributes, this needs to be run on the vertex shader, which means the output has to be run through a vertex interpolator. The auto-generated template material creates an unattached vertex interpolator next to feature tables using per-vertex feature id attributes, as a reminder.
autoFill2

##### Add the material layer to to a material instance, as usual.
autoFill3

##### The final output: the NYC buildings tileset, with all buildings from 1931 highlighted yellow.
autoFill4

### CODE REVIEW

**Comment by cesium-concierge** [comment]
Thanks for the pull request @nithinp7!

* :heavy_check_mark: Signed CLA found.
* :grey_question: [CHANGES.md](https://github.com/CesiumGS/cesium-unreal/blob/metadata-textures/CHANGES.md) was not updated.
    * If this change updates the public API in any way, please add a bullet point to `CHANGES.md`.

Reviewers, don't forget to make sure that:

- [ ] [Cesium for Unreal Samples](https://github.com/CesiumGS/cesium-unreal-samples) works.

**Comment by kring** [comment]
@nithinp7 this looks really exciting! Seems like you took it much farther than I realized. I think it would be helpful if you could walk me through it next week. And can you share the blueprint for the custom materials you've demonstrated in the screenshots above?

**Comment by nithinp7** [comment]
Just added an example material that can be added to the "New York City 3D Buildings" tileset to color based on height. The material and the material layer should be under "Cesium for Unreal Content -> Materials -> Experimental".

File: Source/CesiumRuntime/Public/CesiumMetadataModel.h [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+public:
+  FCesiumMetadataModel() {}
+
+  FCesiumMetadataModel(
+      const CesiumGltf::Model& model,
+      const CesiumGltf::ExtensionModelExtFeatureMetadata& metadata);
+
+private:
+  TMap<FString, FCesiumMetadataFeatureTable> _featureTables;
+  TMap<FString, FCesiumFeatureTexture> _featureTextures;
```
Comment (nithinp7) [inline Source/CesiumRuntime/Public/CesiumMetadataModel.h:31]: @kring Following up from offline discussion, these are the only "duplicated" metadata that are being stored as part of the updates to the metadata blueprint api. Particularly the duplication is just the string keys; the mapped values are wrapped CesiumGltf views. In FCesiumMetadataFeatureTable a similar TMap already exists (from before this PR) with string keys for property names, but again the mapped property arrays are not being duplicated, FCesiumMetadataProperty still wraps a CesiumGltf view into property arrays just as before.

**Comment by nithinp7** [comment]
@kring Officially marking this as ready to review! I just added the `UCesiumEncodedMetadataComponent` which can be added to a tileset. The list of feature table properties and feature textures to encode can be described in the details panel of the component. Later, only matching properties will be encoded and usable in the material. I still need to make the tileset automatically pick up changes in the component.

**Comment by kring** [comment]
The two new uassets were created in UE 4.27, so they can't be loaded in UE 4.26.

**Comment by kring** [comment]
This is amazing! I think there are a few things we can tweak to improve the usability though:
* Add doc comments for `FPropertyDescription` etc (you already have a TODO). Also the Auto Fill button.
* Designate a `TitleProperty` for the array of properties. Probably the property name. That way users can which property is which without expanding them. See the CesiumSubLevels property in CesiumGeoreference.h for an example.
* I think the way Auto Fill works is a little painful, where it adds all the properties and those have a cost even if you don't use them. And avoiding that cost requires deleting the properties, which could be super tedious, especially with a tileset with lots of properties (Cesium OSM Buildings, for example, could easily have a hundred or more properties just from the tiles visible on screen). I think it would be a drastic improvement - and hopefully not too hard - to make Auto Fill more of an "Add" button that opens a dialog to select the properties to add.
* Especially with the above, I wonder if a flat structure where "feature table name" is one of the fields might be easier to work with than the tree structure. So basically CesiumEncodedMetadata would have a simple list of properties with these attributes:
  * Feature Table Name
  * Property Name
  * Component Type
  * Type
  * Normalized
  * (Maybe!)
* Not really usability, but... Should the texture sampler type used for the feature material be "Data" instead of "Color"? I'm admittedly not real clear on the difference, but I think the Color sampler might do sRGB correction or somesuch, which seems bad. Or maybe that doesn't happen because of the PropertyArray.Load?
* Would love to be able to create these kinds of materials without writing HLSL. I guess your other branch is the ultimate answer there?

**Comment by kring** [comment]
I'm probably doing something dumb, but I can't seem to get this to work. I get strange compiler errors from my material instance:
image

My metadata extension looks like this:
image

Material Layer (`ML_Material`) looks like this:
image

And material instance looks like this:
image

Any idea what I'm missing? This is with the NYC Buildings tileset.

Also I noticed that the picking blueprint in 06_CesiumMetadata doesn't work any more, presumably because changes in the API mean it doesn't compile anymore. In the short term, do you have an updated version of the blueprint? But before we ship this, we should make sure we preserve backward compatibility if possible.

**Comment by nithinp7** [comment]
@kring I have been running into that issue with per-vertex feature id attributes only. I think it is due to the vertex interpolator, the vertex interpolator forces the input graph to compile on the vertex shader which is needed here. The only work-around I've found so far has been to remove all layers below the metadata layer (so this is only usable if _none_ of the original glTF shading is needed in the metadata styling, e.g. metadata styled tints wouldn't be possible). I replaced the background layer with the "Empty" material.

**Comment by kring** [comment]
Ok if I delete the extra material layers and use ML_CesiumNothing as the base, the compile errors go away. It still doesn't work the way I expect, though. I'm testing with NYC Buildings, asset ID 75343. The problem is that, even though I've tried to arrange for every building with a height to be red, most of the buildings are black. I can see by exploring this tileset in CesiumJS that they all (or at least most) have a valid Height property.

image

My material layer, ML_Material, looks like this:

image

The encoded metadata extension is straightforward:

image

And the material instance looks like this:

image

I noticed that you used the Empire State Building in your example above, and that happens to be one of the "working" buildings. Is there perhaps some problem where most of the buildings are missing their metadata?

**Comment by kring** [comment]
I haven't reviewed the code in great detail, but here are the things I think need to happen from a user-facing perspective before we can ship this:

* [ ] Assets need to target UE 4.26, not 4.27.
* [ ] `06_Metadata` level blueprint has a bunch of errors. We need a backward compatibility path. Should be possible with some Core Redirects.
* [ ] Generated materials should have a better name, probably involving tileset name and the names of the properties included in the material.
* [ ] Can we open the content browser and auto-select the newly-generated material? Otherwise it's easy to miss that the "Generate Material" button did anything.
* [ ] Is it possible to only rewrite the generated portion of an existing material? Like the set of textures and the custom expression node, but not anything attached to it?
* [ ] If not, we should at least warn before overwriting an existing generated material.
* [ ] Those compiler errors when the material is too complicated (?): We at least need to understand what's going on there well enough to explain it to users. Ideally we'd fix it somehow, if possible.

**Comment by nithinp7** [comment]
@kring Concerning two of those points, maybe it would be a good idea to enumerate the saved material layers to avoid a name clash and overwrite. I'll look for a way to scan the directory for existing materials of that name and tack on a number to deconflict.

**Comment by kring** [comment]
> I'll look for a way to scan the directory for existing materials of that name and tack on a number to deconflict.

Yeah that'd definitely be an improvement! Not quite as good as only regenerating the generated portion, of course.

**Review by nithinp7 — COMMENTED** [review]
An exhaustive list of breaking changes to the metadata blueprint API

**Comment by kring** [comment]
You might already be aware, but `MLB_CesiumAlphaBlendBaseColor.uasset` has been upgraded to 4.27. It needs to go back to 4.26.

**Comment by nithinp7** [comment]
@kring That was quite painful to sort out, but the blueprint api changes are now backwards compatible with deprecation warnings as necessary! I will open a follow-up PR to remove the deprecated mess.

Remaining tasks: 
- Make the example materials 4.26-compatible and undo the accidental change to the material blend. 
- Investigate vertex-interpolator issue.
- Add button on the EncodedMetadataComponent to update the tileset to use the metadata selection (essentially a more convenient way to just refresh the tileset). I was going to have the tileset automatically pick up changes, but this might be extremely costly the way Auto-Fill currently works. _All_ of the metadata would be loaded to the GPU if we refresh the tileset before the user deletes unneeded metadata.
- Regenerate the generated portion of materials (I'm going to time-box this and bail if it takes too long).

**Comment by nithinp7** [comment]
@kring Alright another somewhat painful task, but I think the regenerating nodes idea is implemented reasonably now! There is a new "Target Material Layer" field on the encoded metadata component, that indicates what material layer file was just generated / what file to regenerate. This makes the regeneration robust to moving the material layer into a different directory and also allows users to pick which specific file they want to regenerate.

A few more quality-of-life features came out of this:
- The material editor automatically opens to a newly generated / regenerated material.
- Any existing editor window of the material layer with unsaved changes is prompted to save before a material regeneration actually happens.
- Generated / regenerated materials are now automatically saved.
- If the "Target Material Layer" is empty, a new material layer is generated in the /Game/ folder and the "Target Material Layer" is automatically set to the newly created layer so that it is the immediate target of future regenerations.

This unfortunately introduces a new plugin dependency to be able to save the resulting material (although I was hoping this is a built-in plugin like Water, SunSky, etc). I _think_ the saving was necessary to properly refresh what the material editor showed when opening up regenerated materials. Previously it was showing the out-of-date version in the material editor until the material was saved and the material editor was re-opened. This new dependency is causing CI to fail I believe, maybe we need to add the plugin to the version of Unreal we have on s3?

**Comment by nithinp7** [comment]
Added two example metadata styling materials for NYC buildings in the CesiumforUnrealContent/MetadataStyling folder, one visualizing height and the other visualizing year-of-construction.

**Comment by kring** [comment]
🤩 this is looking awesome! I'll do a more thorough review tomorrow, but I tried it out and I think it's really close. Do you think we should try to get it into the April release? My inclination is not to rush it, given that it's a very large new feature. Giving everyone some time to play with it before we ship it is probably worthwhile. What do you think?

File: Source/CesiumRuntime/Public/CesiumFeatureIdAttribute.h [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+#include "Kismet/BlueprintFunctionLibrary.h"
+#include "CesiumFeatureIdAttribute.generated.h"
+
+namespace CesiumGltf {
+struct Model;
+struct Accessor;
+struct FeatureTable;
+} // namespace CesiumGltf
+
+USTRUCT(BlueprintType)
```
Comment (kring) [inline Source/CesiumRuntime/Public/CesiumFeatureIdAttribute.h:16]: This file needs doc.

File: Source/CesiumRuntime/Public/CesiumFeatureIdTexture.h [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+#include "Containers/UnrealString.h"
+#include "Kismet/BlueprintFunctionLibrary.h"
+#include "CesiumFeatureIdTexture.generated.h"
+
+namespace CesiumGltf {
+struct Model;
+struct FeatureIDTexture;
+} // namespace CesiumGltf
+
+USTRUCT(BlueprintType)
```
Comment (kring) [inline Source/CesiumRuntime/Public/CesiumFeatureIdTexture.h:16]: Doc here too.

File: Source/CesiumRuntime/Public/CesiumFeatureTextureProperty.h [hunk 4]
```diff
@@ -0,0 +1,113 @@
+#pragma once
+
+#include "CesiumGltf/FeatureTexturePropertyView.h"
+#include "GenericPlatform/GenericPlatform.h"
+#include "Kismet/BlueprintFunctionLibrary.h"
+#include "CesiumFeatureTextureProperty.generated.h"
+
+USTRUCT(BlueprintType)
```
Comment (kring) [inline Source/CesiumRuntime/Public/CesiumFeatureTextureProperty.h:8]: Doc in this file.

File: Source/CesiumRuntime/Public/CesiumFeatureTexture.h [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+#include "Containers/Array.h"
+#include "Kismet/BlueprintFunctionLibrary.h"
+#include "CesiumFeatureTexture.generated.h"
+
+namespace CesiumGltf {
+struct Model;
+struct FeatureTexture;
+}; // namespace CesiumGltf
+
+USTRUCT(BlueprintType)
```
Comment (kring) [inline Source/CesiumRuntime/Public/CesiumFeatureTexture.h:14]: doc

File: Source/CesiumRuntime/Public/CesiumMetadataModel.h [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+#include "Containers/UnrealString.h"
+
+#include "CesiumMetadataModel.generated.h"
+
+namespace CesiumGltf {
+struct ExtensionModelExtFeatureMetadata;
+struct Model;
+} // namespace CesiumGltf
+
+USTRUCT(BlueprintType)
```
Comment (kring) [inline Source/CesiumRuntime/Public/CesiumMetadataModel.h:18]: doc

File: Source/CesiumRuntime/Public/CesiumMetadataPrimitive.h [hunk 7]
```diff
@@ -37,21 +46,24 @@ struct CESIUMRUNTIME_API FCesiumMetadataPrimitive {
...
       const CesiumGltf::MeshPrimitive& primitive,
-      const CesiumGltf::ExtensionModelExtFeatureMetadata& metadata,
-      const CesiumGltf::ExtensionMeshPrimitiveExtFeatureMetadata&
-          primitiveMetadata);
+      const CesiumGltf::ExtensionMeshPrimitiveExtFeatureMetadata& metadata,
+      const CesiumGltf::ExtensionModelExtFeatureMetadata& modelMetadata);
 
 private:
-  TArray<FCesiumMetadataFeatureTable> _featureTables;
+  TArray<FCesiumFeatureIdAttribute> _featureIdAttributes;
```
Comment (kring) [inline Source/CesiumRuntime/Public/CesiumMetadataPrimitive.h:59]: The doc for the `FCesiumMetadataPrimitive` class says "It holds a collection of feature tables." That's not true anymore and should be updated.

File: Source/CesiumRuntime/Public/CesiumMetadataPrimitive.h [hunk 8]
```diff
@@ -67,11 +79,50 @@ class CESIUMRUNTIME_API UCesiumMetadataPrimitiveBlueprintLibrary
...
+      Meta =
+          (DeprecatedFunction,
+           DeprecationMessage =
+               "Use UCesiumMetadataModelBlueprintLibrary::GetFeatureTables() instead."))
   static const TArray<FCesiumMetadataFeatureTable>&
   GetFeatureTables(UPARAM(ref)
                        const FCesiumMetadataPrimitive& MetadataPrimitive);
 
+  /**
+   * @brief Get all the vertex features that are associated with the primitive.
```
Comment (kring) [inline Source/CesiumRuntime/Public/CesiumMetadataPrimitive.h:92]: "vertex features"? That doesn't sound right.

File: Source/CesiumRuntime/Public/CesiumMetadataValueType.h [hunk 9]
```diff
@@ -20,6 +20,11 @@ enum class ECesiumMetadataBlueprintType : uint8 {
   Array
 };
 
+// True types are cast, reintepreted, or parsed before being packed into gpu
+// types when encoding into a texture.
```
Comment (kring) [inline Source/CesiumRuntime/Public/CesiumMetadataValueType.h:24]: I think this isn't the right comment style to make it show up as help text in Blueprints. It could probably be a bit more user-facing too.

**Review by kring — COMMENTED** [review]
Awesome work @nithinp7, this was a huge effort! I have a bunch of comments below, mostly just small things.

I updated the metadata level in cesium-unreal-samples to use the non-deprecated APIs. It's a little messy, but here's what I came up with. Just posting it here cause it might be useful when we go to update the samples.

image

Or on blueprintue:
https://blueprintue.com/blueprint/qxnxb46i/

File: Source/CesiumRuntime/Private/CesiumEncodedMetadataUtility.cpp [hunk 10]
```diff
@@ -0,0 +1,707 @@
...
+bool encodeMetadataFeatureTableGameThreadPart(
+    EncodedMetadataFeatureTable& encodedFeatureTable) {
+  bool success = true;
+
+  for (EncodedMetadataProperty& encodedProperty :
+       encodedFeatureTable.encodedProperties) {
+    success &=
+        loadTextureGameThreadPart(encodedProperty.pTexture.Get()) != nullptr;
+    // TODO: check if this is the safest way
+    encodedProperty.pTexture->pTexture->AddToRoot();
```
[the remark this answers is not in the document]
Comment (nithinp7) [reply Source/CesiumRuntime/Private/CesiumEncodedMetadataUtility.cpp:644]: @kring At some point there was a crash related to destruction of gltf-level textures held on the gltf component (e.g., feature tables and feature textures that are referenced from the materials of several primitives). I _thought_ this had been the fix, but I just commented out the AddToRoot and RemoveFromRoot calls and it does not seem to crash in any of my testing. I'll remove them and keep and eye out for that crash.

File: Source/CesiumRuntime/Private/CesiumGltfComponent.cpp [hunk 11]
```diff
@@ -1666,6 +1986,7 @@ static void loadPrimitiveGameThreadPart(
   // mesh or not. We don't want the editor creating collision meshes itself in
   // the game thread, because that would be slow.
   pBodySetup->bCreatedPhysicsMeshes = true;
+  pBodySetup->bSupportUVsAndFaceRemap = true;
```
[the remark this answers is not in the document]
Comment (nithinp7) [reply Source/CesiumRuntime/Private/CesiumGltfComponent.cpp:1989]: I think users have to enable UV results from line traces as a project wide setting and then it shows up as part of the HitResult struct. If the above option is set to true and if the project wide setting is set to true, then I think the physX cooking saves the UV coordinates for line traces later. This will take up more CPU memory, but I doubt it should affect cook time (UVs are not needed for collision detection, so BVH insertion or whatever spatial pre-processing is happening will probably not do anything special with the UVs). Later line traces will have to do a simple UV interpolation, but that is trivial as well. 

So TLDR; the cost is just extra CPU memory for the physX mesh.

File: Source/CesiumRuntime/Private/CesiumMetadataGenericValue.cpp [hunk 12]
```diff
@@ -104,9 +104,16 @@ FCesiumMetadataArray UCesiumMetadataGenericValueBlueprintLibrary::GetArray(
...
   return std::visit(
       [](auto value) -> FCesiumMetadataArray {
-        return CesiumMetadataConversions<
-            FCesiumMetadataArray,
-            decltype(value)>::convert(value, FCesiumMetadataArray());
+        auto createArrayView = [](const auto& array) -> FCesiumMetadataArray {
+          return FCesiumMetadataArray(array);
+        };
+
+        // TODO: Is this the best way?
```
[the remark this answers is not in the document]
Comment (nithinp7) [reply Source/CesiumRuntime/Private/CesiumMetadataGenericValue.cpp:111]: @kring I don't think the conversion system handled arrays, but it also didn't make much sense to define implicit conversions between arrays and non-array types. The old code crashed when trying to GetArray I believe. So I check if it is an array type and return an empty array otherwise. The array's component types can still be coerced using the convert code. Maybe this should be paired with an "IsArray" BP function.

That being said, the templating stuff always gets me a bit confused, so if you have a better idea let me know. Particularly if you think conversions of non-array types to array and vice-versa would make sense.

File: Source/CesiumRuntime/Public/Cesium3DTileset.h [hunk 13]
```diff
@@ -819,6 +820,8 @@ class CESIUMRUNTIME_API ACesium3DTileset : public AActor {
 private:
   TUniquePtr<Cesium3DTilesSelection::Tileset> _pTileset;
 
+  FMetadataDescription encodedMetadataDescription;
```
[the remark this answers is not in the document]
Comment (nithinp7) [reply Source/CesiumRuntime/Public/Cesium3DTileset.h:823]: @kring I think that's how I implemented it initially but I don't think it's safe to access components in the load thread, so at the minimum the metadata description needs to be copied over to the tileset on LoadTileset. Accessing a description housed on the tileset from the load thread is thread-safe because the tileset destructor waits for all async tasks to finish before destruction. So all references to the metadata description, from the gltf and primitive components will be gone before the tileset along with the description are deleted.

Maybe there could be a better API to retrieve the description from the tileset on any thread, so that the resource preparer does not need to be friends. Would that be better?

File: Source/CesiumRuntime/Public/CesiumMetadataValueType.h (hunk 9 above)
Comment (nithinp7) [reply Source/CesiumRuntime/Public/CesiumMetadataValueType.h:24]: This struct wasn't intended to be BP-accessible, it isn't exposed anywhere, I removed the UENUM and BlueprintType specifier. There is an almost equivalent version of the enum in the CesiumEncodedMetadataComponent, that one is exposed and documented properly.

File: Source/CesiumRuntime/Private/CesiumGltfComponent.cpp [hunk 14]
```diff
@@ -1558,6 +1862,8 @@ static void loadPrimitiveGameThreadPart(
   pMaterial->SetFlags(
       RF_Transient | RF_DuplicateTransient | RF_TextExportTransient);
 
+  // TODO: why were these called twice, is it okay to remove?
```
[the remark this answers is not in the document]
Comment (nithinp7) [reply Source/CesiumRuntime/Private/CesiumGltfComponent.cpp:1863]: Could we check whether material layers are even used and do it the old way if there are no layers?

[the remark this answers is not in the document]
Comment (nithinp7) [reply Source/CesiumRuntime/Private/CesiumGltfComponent.cpp:1863]: The only concern I had was vaguely that it would pollute some "global namespace" between all the layers, which might particularly effect the metadata layer that can have all sorts of custom names. Admittedly it's contrived, but it brought my attention to these lines. If that's not a worry I'll leave it as is, uncommenting it for now.

File: Source/CesiumRuntime/Private/CesiumEncodedMetadataUtility.cpp [hunk 15]
```diff
@@ -0,0 +1,707 @@
...
+
+    if (!encodedProperty.pTexture->pTextureData) {
+      // TODO: print error?
+      break;
+    }
+
+    FTexture2DMipMap* pMip = new FTexture2DMipMap();
+    encodedProperty.pTexture->pTextureData->Mips.Add(pMip);
+    pMip->SizeX = propertyArraySize;
+    pMip->SizeY = 1;
```
[the remark this answers is not in the document]
Comment (nithinp7) [reply Source/CesiumRuntime/Private/CesiumEncodedMetadataUtility.cpp:213]: @kring That's a great point! Fortunately implementing something like that shouldn't change the encoding code at all since the written buffer will still be linear. It will change what we set the mip->SizeX and mip->SizeY to and also how we retrieve the property in the auto-generated material (i.e., the propertyArray.Load). 

What would be a good max row / max column size (in pixels)?

File: Source/CesiumRuntime/Private/CesiumGltfComponent.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (kring) [reply Source/CesiumRuntime/Private/CesiumGltfComponent.cpp]: I don't think setting global parameters will affect layer parameters at all, even if they share a name. But setting this parameter values that probably don't even exist is a bit wasteful, and it does prevent users from using the same names for global (non-layer) parameters. I wrote #814 for this.

File: Source/CesiumRuntime/Private/CesiumMetadataGenericValue.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (kring) [reply Source/CesiumRuntime/Private/CesiumMetadataGenericValue.cpp:111]: Ah yeah looks like the old code would have just returned an empty `FMetadataArray` in all cases. I think it does make sense to convert any type to an array with one element of that type, but we can add that if/when we need it.

File: Source/CesiumRuntime/Public/Cesium3DTileset.h (hunk 13 above)
[the remark this answers is not in the document]
Comment (kring) [reply Source/CesiumRuntime/Public/Cesium3DTileset.h]: Anytime we find ourselves making case a like "this is thread-safe because..." we should stop and consider whether we're approaching this correctly. Because:

1. Proving thread-safety is hard. It's easy to miss something.
2. Seemingly innocent changes later can invalidate the argument that enabled us to convince ourselves it was thread-safe. We might not realize this invalidation has happened and end up creating an intermittent and hard to debug failure, maybe without even realizing it until months later.

For example, it's not enough to be sure the lifetime of the object you're accessing from the worker thread will be long enough. We also have to be sure that that object won't change at all. Because any change in one thread while accessing from another thread can (unless we're extremely careful) lead to undefined behavior.

I agree with you in this case that it's safe. Still, why rely on that unless we have to? In this case the problem isn't that copying for the worker thread is expensive (it's not); the problem is how to copy it to the worker thread at all? I think the answer to that is to use the same pattern we just adopted for raster overlays, where `prepareRasterInLoadThread` takes a `rendererOptions` parameter that is set on RasterOverlayOptions and then passed by copy to the worker thread. We should do the same thing with `prepareInLoadThread` and `TilesetOptions`.  Then `alwaysIncludeTangents` can go in the renderer options as well, and we completely eliminate the need to access `this` from the worker.

[the remark this answers is not in the document]
Comment (kring) [reply Source/CesiumRuntime/Public/Cesium3DTileset.h]: I wrote https://github.com/CesiumGS/cesium-native/issues/480, it doesn't have to be part of this PR.

**Comment by kring** [comment]
This looks great, I think the only thing stopping me from hitting merge is that change we talked about to use square-ish textures for metadata so that our feature count isn't limited to 4096 or 8192 or whatever the max texture size is.

**Comment by nithinp7** [comment]
@kring Should be good to go now, changed the feature table encoding and material generation code to create and use square textures. Also made slight changes so the material editor for new / edited materials always gets focused when it is opened. Lastly, found a way to mark the material compilation as dirty in the editor, so the green "Apply" button shows up after regeneration. I never noticed until today that the "Apply" was greyed out after regeneration (maybe it was a regression from the last volley of commits), but it was making it difficult to recompile without making an extraneous change to get the editor's attention.

Anyways, it should be simple to apply the square-texture fix to any previously-generated material layers,: Just select it as the target material, check that the same properties are selected, and regenerate.

**Comment by nithinp7** [comment]
@kring As an aside, I'm super glad you suggested and pushed me to implement the regeneration, it makes the workflow very flexible!

**Comment by kring** [comment]
I spent some time trying to understand those shader compilation errors when using `ML_CesiumGltf` as the background (bottom) layer. Here's what I understand to be happening...

Unreal Engine and/or the hardware have a limit of 8 "custom UV" interpolators, each of which is two floats. Older shader models have less, but SM5, Vulkan, and Metal (which are the only ones we care about, really) all seem to allow at least 8. These 16 floats are the only way to get information from the vertex to the pixel shader. So regular UV texture coordinates coordinates start out as vertex attributes, and if you want to use them from the pixel shader side of the house, then congrats, you just used 2 of your 16 available interpolator slots.

Unreal Engine materials aren't super obvious about which parts of a material are vertex shader and which are pixel shader. From what I understand, it comes down to the material output pins. Pretty much all of the material output pins imply pixel shader. So if you connect nodes to "Base Color" for example, everything feeding into that runs at pixel shader time. The exception is special "Customized UVn" pins. Nodes connected to those run in the vertex shader. A common trick in older versions of UE was to do some per-vertex work, connect the output to Customized UVn pins as necessary, and then have separate nodes that read TexCoord[n], do some further computations in the pixel shader, and send the output to normal pins like Base Color. In more recent versions of UE, we can use "Vertex Interpolator" nodes instead. A Vertex Interpolator node has a "VS" input and a "PS" output. Nodes attached to the VS side run in the vertex shader, the value(s) attached to the VS input are stored in Customized UV channels as necessary, and nodes attached to the PS side use those values to do further computation in the pixel shader. The release notes for UE4.16 have a better explanation of this than mine: https://docs.unrealengine.com/4.26/en-US/WhatsNew/Builds/ReleaseNotes/4_16/#new:vertexinterpolatormaterialexpressionnode

So Vertex Interpolators are really handy for passing values between the vertex and pixel shader, because we don't have to manually allocate Customized UVn slots to them. But it's important to realize that UE is still allocating them behind the scenes. In order to do this, it looks at which TexCoord[n] values are used from the pixel shader. Those need to be assigned directly to equivalent Customized UVn pins, otherwise the pixel shader wouldn't be able to use them. Then values passed through vertex interpolators are assigned to further Customized UVn pins, automatically and transparently.

But here is where we run into trouble. Because glTF has lots of options, and we need an uber-shader that can handle all of them, our `ML_CesiumGltf` uses every single one of the Customzied UVn slots. It doesn't really, of course. Even if you use every single glTF feature, and separate texture coordinates for each, you'll only use 5 sets of texture coordinates (base color, metallic/roughness, emissive, normal, occlusion). That should leave 3 available. But we actually use all of them because we don't know which ones we're going to use. Every texture sample pulls its coordinates through our `CesiumSelectTexCoords` function, which is basically a giant switch that turns a texture coordinate index into texture coordinates. Because the switch includes all 8 texture coordinates, our M_CesiumGltf pixel shader effectively uses all 8 Customzied UVn pins. That leaves zero available for the Vertex Interpolator, and so we get (strange, not very helpful) compiler errors.

One improvement we can make right away is to use vertex interpolators in the `ML_CesiumGltf` layer right after each `CesiumSelectTexCoords`. This way each use of `CesiumSelectTexCoords` (of which there are 5) will use up a `Customized UV_n` slot, instead of each possible set of texture coordinates (of which there are 8) using up a slot. So that'll free up 3 UVs / 6 scalars for use with material styling. I think it might give us a tiny performance improvement too, so I'll go ahead and make that change.

But this is going to be really difficult to fix more generally. Ultimately maybe we need a way to generate the glTF base material from representative tiles, so every texture can just use the right glTF materials and texture coordinates and be done with it.

**Comment by kring** [comment]
One thing I'm still not completely clear on is why we have to use a vertex interpolator for metadata styling. As long as the feature IDs are the same for all three vertices in every triangle, it seems like it shouldn't make any difference. Yet there are crazy artifacts without the vertex interpolator.
image

Why is it so terrible to sample metadata in the pixel shader?

**Comment by kring** [comment]
> One improvement we can make right away is to use vertex interpolators in the ML_CesiumGltf layer right after each CesiumSelectTexCoords.

ehhh... the problem with this is that we should do the same for the `CesiumSelectTexCoords` in `MF_CesiumSampleRasterOverlay`. But if we do that, our standard `MI_CesiumThreeOverlaysAndClippingAndWater` doesn't compile anymore. 5 glTF vertex interpolators, plus three overlay vertex interpolators, plus one clipping vertex interpolator is 9 vertex interpolators, one more than we're allowed.

**Comment by kring** [comment]
Ok I figured out the cause of those crazy artifacts: Draco encoding. The feature ID attributes are Draco-compressed, so they're quantized, so they're not exactly integers. So the truncation to integer would sometimes be off by one, and when we sampled per pixel this could cause very different feature values even within a single triangle. The solution is to round the sampled feature ID before truncating them to integers. We had the exact same bug in the original picking implementation, so I'm embarassed I didn't think of it sooner.

Anyway, with that, metadata styling now works well without a Vertex Interpolator. Actually it works better, because you don't have to remember to swap out the ML_CesiumGltf for ML_Nothing. It might be a little slower, but IMO that all falls into the category of "to go faster, make a custom material for your tileset, rather than using the uber-materials we provide." So I removed the vertex interpolator from the material generator.

Before I figured out that Draco was the issue, I thought it might be sRGB colorspace conversion. So I also added a bit of code to only treat the glTF base color and emissive textures as sRGB, and the others as not sRGB (as the glTF spec says we should). It didn't help with the metadata problem at all, but I think it's a good change so I didn't revert it.

So with that I'm really happy with this PR and ready to merge it. But because I made some non-trivial changes at the last minute here, @nithinp7, you should take a look and make sure it all still looks ok to you. If you're happy, go ahead and merge it!

**Comment by nithinp7** [comment]
Thanks for looking into this @kring! I have no objections with any of the changes you made, looks good to me. My only remaining concern, as discussed offline, is to make sure I understand what the draco issue is and why rounding in the shader fixes it. My alternate theory for the artifacts was that all three feature ids on the vertices of the triangle _are_ the same even after the quantized -> integer -> float conversions, but the automatic barycentric interpolation of the three same floats in the pixel shader is still subject to some ever-so-slight floating point error. Just enough, to make the truncation back to `uint propertyIndex` cause an incorrect feature id.

**Comment by nithinp7** [comment]
Self-merging this now with approval from @kring! 🤞
